File: src/main/java/cc/mirukuneko/realtrainmodrenewed/installedobject/InstalledObjectDefinition.java
package cc.mirukuneko.realtrainmodrenewed.installedobject;

import net.minecraft.world.phys.Vec3;

import java.util.List;
import java.util.Map;

public class InstalledObjectDefinition {
    private final String id;
    private final String displayName;
    private final String packName;
    private final InstalledObjectCategory category;
    private final String modelFile;
    private final String scriptPath;
    private final String buttonTexture;
    private final Map<String, String> textureOverrides;
    private final Vec3 modelOffset;
    private final float modelScale;
    private final boolean smoothing;
    private final float width;
    private final float height;
    private final float depth;
    private final String signTexture;
    private final String emissiveTexture;
    private final String runningSound;
    private final Map<Integer, List<String>> signalLightGroups;
    private final List<String> renderObjects;
    private final Vec3 scriptBodyPos;
    private final int signFrame;
    private final int backTexture;
    private Vec3 wireAttachPos = Vec3.ZERO;
    // ワイヤー用パラメータ(WireConfig 相当)。コンストラクタ後に setWireParams で設定。
    private float sectionLength = 0.5F;
    private float deflectionCoefficient = 0.0F;

    public float getSectionLength() {
        return sectionLength;
    }

    public float getDeflectionCoefficient() {
        return deflectionCoefficient;
    }

    public void setWireParams(float sectionLength, float deflectionCoefficient) {
        if (sectionLength > 0.0F) {
            this.sectionLength = sectionLength;
        }
        this.deflectionCoefficient = Math.max(0.0F, deflectionCoefficient);
    }

    public Vec3 getWireAttachPos() {
        return wireAttachPos;
    }

    public void setWireAttachPos(Vec3 wireAttachPos) {
        this.wireAttachPos = wireAttachPos == null ? Vec3.ZERO : wireAttachPos;
    }

    public InstalledObjectDefinition(String id, String displayName, String packName, InstalledObjectCategory category,
                                     String modelFile, String scriptPath, Map<String, String> textureOverrides,
                                     Vec3 modelOffset, float modelScale, boolean smoothing,
                                     float width, float height, float depth, String signTexture) {
        this(id, displayName, packName, category, modelFile, scriptPath, "", textureOverrides, modelOffset, modelScale,
            smoothing, width, height, depth, signTexture, "", "", Map.of(), Vec3.ZERO, 1, 1);
    }

    public InstalledObjectDefinition(String id, String displayName, String packName, InstalledObjectCategory category,
                                     String modelFile, String scriptPath, String buttonTexture, Map<String, String> textureOverrides,
                                     Vec3 modelOffset, float modelScale, boolean smoothing,
                                     float width, float height, float depth, String signTexture,
                                     String emissiveTexture, String runningSound, Map<Integer, List<String>> signalLightGroups,
                                     Vec3 scriptBodyPos) {
        this(id, displayName, packName, category, modelFile, scriptPath, buttonTexture, textureOverrides, modelOffset,
            modelScale, smoothing, width, height, depth, signTexture, emissiveTexture, runningSound,
            signalLightGroups, List.of(), scriptBodyPos, 1, 1);
    }

    public InstalledObjectDefinition(String id, String displayName, String packName, InstalledObjectCategory category,
                                     String modelFile, String scriptPath, String buttonTexture, Map<String, String> textureOverrides,
                                     Vec3 modelOffset, float modelScale, boolean smoothing,
                                     float width, float height, float depth, String signTexture,
                                     String emissiveTexture, String runningSound, Map<Integer, List<String>> signalLightGroups,
                                     Vec3 scriptBodyPos, int signFrame, int backTexture) {
        this(id, displayName, packName, category, modelFile, scriptPath, buttonTexture, textureOverrides, modelOffset,
            modelScale, smoothing, width, height, depth, signTexture, emissiveTexture, runningSound,
            signalLightGroups, List.of(), scriptBodyPos, signFrame, backTexture);
    }

    public InstalledObjectDefinition(String id, String displayName, String packName, InstalledObjectCategory category,
                                     String modelFile, String scriptPath, String buttonTexture, Map<String, String> textureOverrides,
                                     Vec3 modelOffset, float modelScale, boolean smoothing,
                                     float width, float height, float depth, String signTexture,
                                     String emissiveTexture, String runningSound, Map<Integer, List<String>> signalLightGroups,
                                     List<String> renderObjects, Vec3 scriptBodyPos, int signFrame, int backTexture) {
        this.id = id;
        this.displayName = displayName;
        this.packName = packName;
        this.category = category;
        this.modelFile = modelFile;
        this.scriptPath = scriptPath;
        this.buttonTexture = buttonTexture == null ? "" : buttonTexture;
        this.textureOverrides = textureOverrides == null ? Map.of() : Map.copyOf(textureOverrides);
        this.modelOffset = modelOffset == null ? Vec3.ZERO : modelOffset;
        this.modelScale = modelScale <= 0.0F ? 1.0F : modelScale;
        this.smoothing = smoothing;
        this.width = width <= 0.0F ? 1.0F : width;
        this.height = height <= 0.0F ? 1.0F : height;
        this.depth = depth <= 0.0F ? 0.125F : depth;
        this.signTexture = signTexture == null ? "" : signTexture;
        this.emissiveTexture = emissiveTexture == null ? "" : emissiveTexture;
        this.runningSound = runningSound == null ? "" : runningSound;
        this.signalLightGroups = signalLightGroups == null ? Map.of() : Map.copyOf(signalLightGroups);
        this.renderObjects = renderObjects == null ? List.of() : List.copyOf(renderObjects);
        this.scriptBodyPos = scriptBodyPos == null ? Vec3.ZERO : scriptBodyPos;
        this.signFrame = Math.max(1, signFrame);
        this.backTexture = backTexture;
    }

    public String getId() {
        return id;
    }

    public String getDisplayName() {
        return displayName;
    }

    public String getPackName() {
        return packName;
    }

    public InstalledObjectCategory getCategory() {
        return category;
    }

    public String getModelFile() {
        return modelFile;
    }

    public String getScriptPath() {
        return scriptPath;
    }

    public String getButtonTexture() {
        return buttonTexture;
    }

    public Map<String, String> getTextureOverrides() {
        return textureOverrides;
    }

    public Vec3 getModelOffset() {
        return modelOffset;
    }

    public float getModelScale() {
        return modelScale;
    }

    public boolean isSmoothing() {
        return smoothing;
    }

    public float getWidth() {
        return width;
    }

    public float getHeight() {
        return height;
    }

    public float getDepth() {
        return depth;
    }

    public String getSignTexture() {
        return signTexture;
    }

    public String getEmissiveTexture() {
        return emissiveTexture;
    }

    public String getRunningSound() {
        return runningSound;
    }

    public Map<Integer, List<String>> getSignalLightGroups() {
        return signalLightGroups;
    }

    public List<String> getRenderObjects() {
        return renderObjects;
    }

    public int getSignFrame() {
        return signFrame;
    }

    public int getBackTexture() {
        return backTexture;
    }

    public Vec3 getScriptBodyPos() {
        return scriptBodyPos;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/installedobject/InstalledObjectDefinition.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.installedobject

import net.minecraft.world.phys.Vec3
import kotlin.math.max

class InstalledObjectDefinition(
    @JvmField val id: String,
    @JvmField val displayName: String,
    @JvmField val packName: String,
    @JvmField val category: InstalledObjectCategory,
    @JvmField val modelFile: String,
    @JvmField val scriptPath: String,
    buttonTexture: String?,
    textureOverrides: Map<String, String>?,
    modelOffset: Vec3?,
    modelScale: Float,
    @JvmField val smoothing: Boolean,
    width: Float,
    height: Float,
    depth: Float,
    signTexture: String?,
    emissiveTexture: String?,
    runningSound: String?,
    signalLightGroups: Map<Int, List<String>>?,
    renderObjects: List<String>?,
    scriptBodyPos: Vec3?,
    signFrame: Int,
    @JvmField val backTexture: Int
) {
    @JvmField val buttonTexture: String = buttonTexture ?: ""
    @JvmField val textureOverrides: Map<String, String> = textureOverrides ?: emptyMap()
    @JvmField val modelOffset: Vec3 = modelOffset ?: Vec3.ZERO
    @JvmField val modelScale: Float = if (modelScale <= 0.0F) 1.0F else modelScale
    @JvmField val width: Float = if (width <= 0.0F) 1.0F else width
    @JvmField val height: Float = if (height <= 0.0F) 1.0F else height
    @JvmField val depth: Float = if (depth <= 0.0F) 0.125F else depth
    @JvmField val signTexture: String = signTexture ?: ""
    @JvmField val emissiveTexture: String = emissiveTexture ?: ""
    @JvmField val runningSound: String = runningSound ?: ""
    @JvmField val signalLightGroups: Map<Int, List<String>> = signalLightGroups ?: emptyMap()
    @JvmField val renderObjects: List<String> = renderObjects ?: emptyList()
    @JvmField val scriptBodyPos: Vec3 = scriptBodyPos ?: Vec3.ZERO
    @JvmField val signFrame: Int = max(1, signFrame)

    fun getId(): String = id
    fun getDisplayName(): String = displayName
    fun getPackName(): String = packName
    fun getCategory(): InstalledObjectCategory = category
    fun getModelFile(): String = modelFile
    fun getScriptPath(): String = scriptPath
    fun getButtonTexture(): String = buttonTexture
    fun getTextureOverrides(): Map<String, String> = textureOverrides
    fun getModelOffset(): Vec3 = modelOffset
    fun getModelScale(): Float = modelScale
    fun isSmoothing(): Boolean = smoothing
    fun getWidth(): Float = width
    fun getHeight(): Float = height
    fun getDepth(): Float = depth
    fun getSignTexture(): String = signTexture
    fun getEmissiveTexture(): String = emissiveTexture
    fun getRunningSound(): String = runningSound
    fun getSignalLightGroups(): Map<Int, List<String>> = signalLightGroups
    fun getRenderObjects(): List<String> = renderObjects
    fun getScriptBodyPos(): Vec3 = scriptBodyPos
    fun getSignFrame(): Int = signFrame
    fun getBackTexture(): Int = backTexture

    var wireAttachPos: Vec3 = Vec3.ZERO
        private set

    var sectionLength: Float = 0.5F
        private set
    var deflectionCoefficient: Float = 0.0F
        private set

    fun setWireParams(sectionLength: Float, deflectionCoefficient: Float) {
        if (sectionLength > 0.0F) this.sectionLength = sectionLength
        this.deflectionCoefficient = max(0.0F, deflectionCoefficient)
    }

    fun setWireAttachPos(pos: Vec3?) {
        wireAttachPos = pos ?: Vec3.ZERO
    }

    // Convenience constructors matching Java overloads
    constructor(
        id: String, displayName: String, packName: String, category: InstalledObjectCategory,
        modelFile: String, scriptPath: String, textureOverrides: Map<String, String>?,
        modelOffset: Vec3?, modelScale: Float, smoothing: Boolean,
        width: Float, height: Float, depth: Float, signTexture: String?
    ) : this(id, displayName, packName, category, modelFile, scriptPath, "",
        textureOverrides, modelOffset, modelScale, smoothing, width, height, depth, signTexture,
        "", "", emptyMap(), Vec3.ZERO, 1, 1)

    constructor(
        id: String, displayName: String, packName: String, category: InstalledObjectCategory,
        modelFile: String, scriptPath: String, buttonTexture: String?, textureOverrides: Map<String, String>?,
        modelOffset: Vec3?, modelScale: Float, smoothing: Boolean,
        width: Float, height: Float, depth: Float, signTexture: String?,
        emissiveTexture: String?, runningSound: String?, signalLightGroups: Map<Int, List<String>>?,
        scriptBodyPos: Vec3?
    ) : this(id, displayName, packName, category, modelFile, scriptPath, buttonTexture,
        textureOverrides, modelOffset, modelScale, smoothing, width, height, depth, signTexture,
        emissiveTexture, runningSound, signalLightGroups, emptyList(), scriptBodyPos, 1, 1)

    constructor(
        id: String, displayName: String, packName: String, category: InstalledObjectCategory,
        modelFile: String, scriptPath: String, buttonTexture: String?, textureOverrides: Map<String, String>?,
        modelOffset: Vec3?, modelScale: Float, smoothing: Boolean,
        width: Float, height: Float, depth: Float, signTexture: String?,
        emissiveTexture: String?, runningSound: String?, signalLightGroups: Map<Int, List<String>>?,
        scriptBodyPos: Vec3?, signFrame: Int, backTexture: Int
    ) : this(id, displayName, packName, category, modelFile, scriptPath, buttonTexture,
        textureOverrides, modelOffset, modelScale, smoothing, width, height, depth, signTexture,
        emissiveTexture, runningSound, signalLightGroups, emptyList(), scriptBodyPos, signFrame, backTexture)
}
